### PR TITLE
feat: Default scalar field values

### LINKED ISSUE #2952: Support for default values
We should add support for default values

Possible directive for the feature:
```gql
type User {
  name: String
  active: Bool @default(bool: true)
  employer: String @default(string: "ACME")
  age: Int @default(int: 24)
}

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2952

## Description

This PR adds support for default field values using a new `@default` directive.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Added integration tests

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: internal/request/graphql/schema/collection.go [hunk 1]
```diff
@@ -345,6 +369,35 @@ func indexFromAST(directive *ast.Directive) (client.IndexDescription, error) {
 	return desc, nil
 }
 
+func defaultFromAST(
+	field *ast.FieldDefinition,
+	directive *ast.Directive,
+) (any, error) {
+	astNamed, ok := field.Type.(*ast.Named)
+	if !ok {
+		return nil, NewErrDefaultValueNotAllowed(field.Name.Value, field.Type.String())
```
Comment (AndrewSisley) [inline internal/request/graphql/schema/collection.go:378]: todo: Please add tests for this error condition

File: internal/request/graphql/schema/collection.go [hunk 2]
```diff
@@ -345,6 +369,35 @@ func indexFromAST(directive *ast.Directive) (client.IndexDescription, error) {
...
+	}
+	var value any
+	for _, arg := range directive.Arguments {
+		if defaultPropNameToType[arg.Name.Value] != astNamed.Name.Value {
+			return nil, NewErrDefaultValueInvalid(astNamed.Name.Value, arg.Name.Value)
```
Comment (AndrewSisley) [inline internal/request/graphql/schema/collection.go:383]: todo: Please add tests for this error condition

File: client/collection_field_description.go [hunk 3]
```diff
@@ -38,6 +38,9 @@ type CollectionFieldDescription struct {
 	//
 	// Otherwise will be [None].
 	RelationName immutable.Option[string]
+
+	// DefaultValue contains the default value for this field.
+	DefaultValue any
```
Comment (AndrewSisley) [inline client/collection_field_description.go:43]: todo: Please test the PatchCollection behaviour for this new property

Comment (AndrewSisley) [inline client/collection_field_description.go:43]: todo: Please test and document what happens when this is set on a View - and consider adding a validation rule to prevent that (as it makes no sense to me) (and test that, including on Patch).

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
It looks good and is easy to read, but please add some error tests, validation for the default values (in definition_validation.go), View behaviour, and either test PatchCollection works, or temporary block it off (and test that).

File: internal/request/graphql/schema/collection.go (hunk 1 above)
Comment (nasdf) [reply internal/request/graphql/schema/collection.go:378]: done

File: internal/request/graphql/schema/collection.go (hunk 2 above)
Comment (nasdf) [reply internal/request/graphql/schema/collection.go]: done

File: client/collection_field_description.go (hunk 3 above)
Comment (nasdf) [reply client/collection_field_description.go:43]: done

File: tests/integration/mutation/create/with_default_values_test.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			},
+			testUtils.Request{
+				Request: `mutation {
+							create_Users(input: {}) {
+								age
+								active
+								name
+								points
+							}
+						}`,
```
Comment (shahzadlone) [inline tests/integration/mutation/create/with_default_values_test.go]: question/todo: Perhaps document and add another test where the same mutation is requested twice when a unique constraint is there and when duplicate is allowed.

Comment (nasdf) [reply tests/integration/mutation/create/with_default_values_test.go]: done

**Review by shahzadlone — APPROVED** [review]
Cheers thanks for adding the requested tests

File: client/collection_field_description.go (hunk 3 above)
Comment (AndrewSisley) [reply client/collection_field_description.go:43]: Okay cheers, I'm happy with that :) - can you please document that the property is ignored for views on the directive, and on this `CollectionFieldDescription` prop so that users know?

File: tests/integration/mutation/create/with_default_values_test.go [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+			testUtils.SchemaUpdate{
+				Schema: `
+					type Users {
+						name: String @default(string: "Bob")
+						age: Int @default(int: 40)
+					}
+				`,
+			},
+			testUtils.Request{
+				Request: `mutation {
```
Comment (fredcarle) [inline tests/integration/mutation/create/with_default_values_test.go:180]: question: Would it be worth using the `CreateDoc` action instead to test across all clients?

Comment (fredcarle) [reply tests/integration/mutation/create/with_default_values_test.go:180]: This applies to all the integration tests.

**Review by fredcarle — APPROVED** [review]
LGTM. Thanks for all the tests. Lets just resolve my question before merging please.

File: tests/integration/mutation/create/with_default_values_test.go (hunk 5 above)
Comment (AndrewSisley) [reply tests/integration/mutation/create/with_default_values_test.go:180]: Yes, please use the CreateDoc action instead Keenan (nice spot Fred)

File: client/collection_field_description.go (hunk 3 above)
Comment (nasdf) [reply client/collection_field_description.go:45]: done

File: tests/integration/mutation/create/with_default_values_test.go (hunk 5 above)
Comment (nasdf) [reply tests/integration/mutation/create/with_default_values_test.go]: done

File: tests/integration/mutation/create/with_default_values_test.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationCreate_WithDefaultValue_NoValueProvided_CreatedTwice_ReturnsError(t *testing.T) {
+	test := testUtils.TestCase{
+		Description: "Simple create mutation, with default value, no value provided, and created twice",
+		SupportedMutationTypes: immutable.Some([]testUtils.MutationType{
+			// This test will fail if using the collection save
+			// method because it does not create two unique docs
```
Comment (AndrewSisley) [inline tests/integration/mutation/create/with_default_values_test.go:200]: question: I can't spot anything super obvious in the test code as to why `Save` would not work, would you mind expanding a little?

Comment (nasdf) [reply tests/integration/mutation/create/with_default_values_test.go:200]: It seems like calling save does not actually create two unique documents when the fields have the same value. I'm guessing it checks if the docID exists and calls update instead of create.

Comment (AndrewSisley) [reply tests/integration/mutation/create/with_default_values_test.go:200]: Oh true - it does - thanks Keenan.  Might be worth adding that to this code-comment then.

**Review by AndrewSisley — APPROVED** [review]
LGTM, thanks Keenan :)

**Comment by islamaliev** [comment]
Bug bush result: https://github.com/sourcenetwork/defradb/issues/3133
